libc/strftime.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the American National Standards Committee X3, on Information
.\" Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 4, 2007
.Dt STRFTIME 3
.Os Mastodon
.Sh NAME
.Nm strftime 
.Nd format date and time
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In time.h
.Ft size_t
.Fo strftime
.Fa "char * buf"
.Fa "size_t maxsize"
.Fa "const char * format"
.Fa "const struct tm * timeptr"
.Fc
.Sh DESCRIPTION
The
.Fn strftime
function formats the information from
.Fa timeptr
into the buffer
.Fa buf
according to the string pointed to by
.Fa format .
.Pp
The
.Fa format
string consists of zero or more conversion specifications and
ordinary characters.
All ordinary characters are copied directly into the buffer.
A conversion specification consists of a percent sign
.Dq Ql %
and one other character.
.Pp
No more than
.Fa maxsize
characters will be placed into the array.
If the total number of resulting characters, including the terminating
NUL character, is not more than
.Fa maxsize ,
.Fn strftime
returns the number of characters in the array, not counting the
terminating NUL.
Otherwise, zero is returned and the buffer contents are indeterminate.
.Pp
The conversion specifications are copied to the buffer after expansion
as follows:-
.Bl -tag -width "xxxx"
.It Cm \&%A
is replaced by the full weekday name.
.It Cm %a
is replaced by the abbreviated weekday name.
.It Cm \&%B
is replaced by the full month name.
.It Cm %b
is replaced by the abbreviated month name.
.It Cm %C
is replaced by (year / 100) as decimal number; single
digits are preceded by a zero.
.It Cm %c
is replaced by national representation of the date and time
(the format is similar to that produced by
.Xr date 1 ) .
.It Cm \&%D
is equivalent to
.Dq Li %m/%d/%y .
.It Cm %d
is replaced by the day of the month as a decimal number (01-31).
.It Cm %e
is replaced by the day of month as a decimal number (1-31); single
digits are preceded by a blank.
.It Cm \&%F
is equivalent to
.Dq Li %Y-%m-%d .
.It Cm \&%G
is replaced by a year as a decimal number with century.
This year is the one that contains the greater part of
the week (Monday as the first day of the week).
.It Cm %g
is replaced by the same year as in
.Dq Li %G ,
but as a decimal number without century (00-99).
.It Cm \&%H
is replaced by the hour (24-hour clock) as a decimal number (00-23).
.It Cm %h
the same as %b.
.It Cm \&%I
is replaced by the hour (12-hour clock) as a decimal number (01-12).
.It Cm %j
is replaced by the day of the year as a decimal number (001-366).
.It Cm %k
is replaced by the hour (24-hour clock) as a decimal number (0-23);
single digits are preceded by a blank.
.It Cm %l
is replaced by the hour (12-hour clock) as a decimal number (1-12);
single digits are preceded by a blank.
.It Cm \&%M
is replaced by the minute as a decimal number (00-59).
.It Cm %m
is replaced by the month as a decimal number (01-12).
.It Cm %n
is replaced by a newline.
.It Cm %p
is replaced by 
.Dq am
or
.Dq pm
as appropriate.
.It Cm \&%R
is equivalent to
.Dq Li %H:%M .
.It Cm %r
is equivalent to
.Dq Li %I:%M:%S %p .
.It Cm \&%S
is replaced by the second as a decimal number (00-60).
.It Cm %s
is replaced by the number of seconds since the Epoch, UTC (see
.Xr mktime 3 ) .
.It Cm \&%T
is equivalent to
.Dq Li %H:%M:%S .
.It Cm %t
is replaced by a tab.
.It Cm \&%U
is replaced by the week number of the year (Sunday as the first day of
the week) as a decimal number (00-53).
.It Cm %u
is replaced by the weekday (Monday as the first day of the week)
as a decimal number (1-7).
.It Cm \&%V
is replaced by the week number of the year (Monday as the first day of
the week) as a decimal number (01-53).
If the week containing January
1 has four or more days in the new year, then it is week 1; otherwise
it is the last week of the previous year, and the next week is week 1.
.It Cm %v
is equivalent to
.Dq Li %e-%b-%Y .
.It Cm \&%W
is replaced by the week number of the year (Monday as the first day of
the week) as a decimal number (00-53).
.It Cm %w
is replaced by the weekday (Sunday as the first day of the week)
as a decimal number (0-6).
.It Cm \&%X
the same as
.Dq Li \&%T .
.It Cm %x
the same as
.Dq Li \&%D .
.It Cm \&%Y
is replaced by the year with century as a decimal number.
.It Cm %y
is replaced by the year without century as a decimal number (00-99).
.It Cm \&%Z
is replaced by the time zone name.
.It Cm \&%z
is replaced by the time zone offset from UTC; a leading plus sign stands for
east of UTC, a minus sign for west of UTC, hours and minutes follow
with two digits each and no delimiter between them (common form for
RFC 822 date headers).
.It Cm %+
is replaced by national representation of the date and time
(the format is similar to that produced by
.Xr date 1 ) .
.It Cm %%
is replaced by
.Ql % .
.El
.Sh SEE ALSO
.Xr date 1 
.Pp
The peculiar week number and year in the replacements of
.Ql %G ,
.Ql %g
and
.Ql \&%V
are defined in ISO 8601: 1988.
.Sh BUGS
There is no conversion specification for the phase of the moon.


libc/strftime.c
#include <string.h>
#include <time.h>
#include <stdarg.h>

#include <stdio.h>

struct context {
    char *bfr;
    int   bufsiz;
    const struct tm *t;
    int   err;
    int   ix;
};

static void fillftime(struct context *ctx, const char *format);

static void
fillc(struct context *ctx, char c)
{
    if (ctx->ix >= ctx->bufsiz || ctx->ix < 0)
	ctx->err = 1;
    else
	ctx->bfr[ctx->ix++] = c;
}

static void
fills(struct context *ctx, char *s)
{
    int size;


    if (s == 0)
	ctx->err = 1;
    else {
	size = strlen(s);

	if (ctx->ix < 0 || ctx->ix + size >= ctx->bufsiz)
	    ctx->err = 1;
	else {
	    memcpy(ctx->bfr + ctx->ix, s, size);
	    ctx->ix += size;
	}
    }
}

static void
fill(struct context *ctx, const char *fmt, ...)
{
    va_list ptr;
    int written;

    if (ctx->ix < 0)
	ctx->err = 1;
    else {
	va_start(ptr,fmt);

	written = vsnprintf(ctx->bfr+ctx->ix, ctx->bufsiz - ctx->ix, fmt, ptr);

	if (written < 0 || written + ctx->ix > ctx->bufsiz)
	    ctx->err = 1;
	else
	    ctx->ix += written;

	va_end(ptr);
    }
}


static char *
weekday_name(int wday)
{
    switch (wday) {
    case 0: return "Sunday";
    case 1: return "Monday";
    case 2: return "Tuesday";
    case 3: return "Wednesday";
    case 4: return "Thursday";
    case 5: return "Friday";
    case 6: return "Saturday";
    default:return 0;
    }
}

static char *
weekday_abbr(int wday)
{
    switch (wday) {
    case 0: return "Sun";
    case 1: return "Mon";
    case 2: return "Tue";
    case 3: return "Wed";
    case 4: return "Thu";
    case 5: return "Fri";
    case 6: return "Sat";
    default:return 0;
    }
}


static char *
month_name(int month)
{
    switch (month) {
    case 0: return "January";
    case 1: return "February";
    case 2: return "March";
    case 3: return "April";
    case 4: return "May";
    case 5: return "June";
    case 6: return "July";
    case 7: return "August";
    case 8: return "September";
    case 9: return "October";
    case 10:return "November";
    case 11:return "December";
    default: return 0;
    }
}


static char *
month_abbr(int month)
{
    switch (month) {
    case 0: return "Jan";
    case 1: return "Feb";
    case 2: return "Mar";
    case 3: return "Apr";
    case 4: return "May";
    case 5: return "Jun";
    case 6: return "Jul";
    case 7: return "Aug";
    case 8: return "Sep";
    case 9: return "Oct";
    case 10:return "Nov";
    case 11:return "Dec";
    default: return 0;
    }
}


static char *
ampm(int hour)
{
    return (hour < 12) ? "am" : "pm";
}


static int
weekofyear(const struct tm *time, int kind)
{
    int days, weeks;

    switch (kind) {
    default: days = time->tm_wday; break;
    case 1:  days = time->tm_wday-1; break;
    case 2:  days = (time->tm_wday < 3) ? (time->tm_wday-1)
					: (time->tm_wday + 6);
    }


    weeks = ( 6 + time->tm_yday - time->tm_wday )/ 7;

    if (weeks && (days > time->tm_yday % 7)) --weeks;

    return weeks ? weeks : 53/*we're the stub of the last year*/;
}


static void
expand(struct context *ctx, char control)
{
    switch (control) {
    case 0:     ctx->err = 1; break;
    case 'A':   fills(ctx,weekday_name(ctx->t->tm_wday)); break;
    case 'a':   fills(ctx,weekday_abbr(ctx->t->tm_wday)); break;
    case 'B':   fills(ctx,month_name(ctx->t->tm_mon)); break;
    case 'b':   fills(ctx,month_abbr(ctx->t->tm_mon)); break;
    case 'C':   fill(ctx,"%02d", (1900+ctx->t->tm_year) / 100); break;
    case 'c':   expand(ctx, '+'); break;
    case 'D':   fillftime(ctx, "%m/%d/%y"); break;
    case 'd':   fill(ctx,"%02d", ctx->t->tm_mday); break;
    case 'e':   fill(ctx,"%2d", ctx->t->tm_mday); break;
    case 'F':	fillftime(ctx, "%Y-%m-%d"); break;
    case 'g':
    case 'G':   {   int year;

		    year = ctx->t->tm_year + 1900;
		    if (ctx->t->tm_yday <= ctx->t->tm_wday)
			year--;
		    if (control == 'g')
			fill(ctx, "%02d", year%100);
		    else
			fill(ctx, "%d", year);
		}
		break;
    case 'h':   expand(ctx, 'b'); break;
    case 'H':   fill(ctx,"%02d", ctx->t->tm_hour); break;
    case 'I':   fill(ctx,"%02d", ctx->t->tm_hour%12); break;
    case 'j':   fill(ctx,"%03d", ctx->t->tm_yday); break;
    case 'k':   fill(ctx,"%2d", ctx->t->tm_hour); break;
    case 'l':   fill(ctx,"%2d", ctx->t->tm_hour % 12); break;
    case 'M':   fill(ctx,"%02d", ctx->t->tm_min); break;
    case 'm':   fill(ctx,"%02d", ctx->t->tm_mon + 1); break;
    case 'n':   fillc(ctx,'\n'); break;
    case 'p':   fills(ctx,ampm(ctx->t->tm_hour)); break;
    case 'R':   fillftime(ctx, "%H:%M"); break; break;
    case 'r':	fillftime(ctx, "%I:%M:%S %p"); break;
    case 'S':   fill(ctx,"%02d", ctx->t->tm_sec); break;
    case 's':   fill(ctx,"%d", mktime((struct tm*)ctx->t)); break;
    case 'T':   fillftime(ctx, "%H:%M:%S"); break;
    case 't':   fillc(ctx,'\t'); break;
    case 'U':   fill(ctx, "%02d", weekofyear(ctx->t, 0)); break;
    case 'u':   fill(ctx,"%d", ctx->t->tm_wday == 0 ? 7 : ctx->t->tm_wday);
			    break;
    case 'V':	fill(ctx, "%02d", weekofyear(ctx->t,2)); break;
    case 'v':   fillftime(ctx, "%e-%b-%Y"); break;
    case 'W':   fill(ctx, "%02d", weekofyear(ctx->t,1)); break;
    case 'w':   fill(ctx,"%d", ctx->t->tm_wday); break;
    case 'X':   expand(ctx, 'T'); break;
    case 'x':   expand(ctx, 'D'); break;
    case 'Y':   fill(ctx,"%d", ctx->t->tm_year + 1900); break;
    case 'y':   fill(ctx,"%02d", ctx->t->tm_year % 100); break;
    case 'Z':   fills(ctx,tzname[ctx->t->tm_isdst ? 1 : 0]); break;
    case 'z':   {   int plus, tz;
    #if (_LINUX_C_LIB_VERSION_MAJOR == 4) || (_LINUX_C_LIB_VERSION_MAJOR == 5)
		    tz = timezone / 60;
    #else
		    tz = ctx->t->tm_gmtoff / 60;
    #endif
		    plus = ( tz>=0 ) ? '+' : '-';
		    tz %= (12*60);
		    fill(ctx,"%c%02d%02d", plus, abs(tz/60), abs(tz%60));
		}
		break;
    case '+':   fillftime(ctx, "%a %b %d %R:%S %Z %Y"); break;
    default:
    case '%':   fillc(ctx, control); break;
    }
}


static void
fillftime(struct context *ctx, const char *format)
{
    const char *p;

    for (p = format; *p && !ctx->err; ++p) {
	if (*p == '%')
	    expand(ctx, *++p);
	else 
	    fillc(ctx, *p);
    }
}


size_t
strftime(char *res, size_t size, const char *format, const struct tm *time)
{
    struct context c = { res, size, time };

    if ( res == 0 || size <= 0 || time == 0 || format == 0 )
	return 0;

    fillftime(&c, format);
    fillc(&c, 0);
    return c.err ? 0 : (c.ix-1);
}
